File: leetcode-all/剑指Offer09-用两个栈实现队列.py
# 栈是先进后出，队列是先进先出；输入栈，一个输出栈，数据从输入栈到输出栈，从输出栈输出的元素就满足先进先出。
class CQueue:
    stack1 = []
    stack2 = []

    def __init__(self):
        self.stack1 = []
        self.stack2 = []


    def appendTail(self, value: int) -> None:
        self.stack1.append(value)


    def deleteHead_example(self) -> int:
        if self.stack2:
            return self.stack2.pop()
        if not self.stack1:
            return -1
        while self.stack1:
            self.stack2.append(self.stack1)
        return self.stack2.pop()
    
    
    def deleteHead(self) -> int:
        if len(self.stack2) == 0:
            n = len(self.stack1)
            for i in range(n):
                self.stack2.append(self.stack1.pop())
        if len(self.stack2) == 0:
            return -1
        else:
            return self.stack2.pop()


if __name__=="__main__":
    obj = CQueue()
    obj.appendTail(5)
    obj.appendTail(2)
    param_2 = obj.deleteHead()
    print(param_2)

File: leetcode-all/剑指 Offer 09-用两个栈实现队列.js
var CQueue = function(){
    // 为了实现先进先出，每次弹出时若尾栈为空，则将原栈栈中元素弹出后压入尾栈即可；否则尾栈直接弹出栈尾元素即可。
    this.tail = [];
    this.head = [];
};


CQueue.prototype.appendTail = function(value) {
    this.tail.push(value);
};

/**
 * @return {number}
 */
CQueue.prototype.deleteHead = function() {
    if(this.head.length == 0) {
        var i = this.tail.length;
        if(i == 0) {
            return -1;
        }
        while(i != 0) {
            this.head.push(this.tail.pop());
            i--;
        }
    }
    return this.head.pop();
};

/**
 * Your CQueue object will be instantiated and called as such:
 * var obj = new CQueue()
 * obj.appendTail(value)
 * var param_2 = obj.deleteHead()
 */

var obj = new CQueue()
obj.appendTail(3)
console.log(obj.deleteHead());
console.log(obj.deleteHead());
